nifi-nar-bundles/nifi-graph-bundle/nifi-graph-processors/src/main/java/org/apache/nifi/processors/graph/ExecuteGraphQuery.java

```diff
@@ -0,0 +1,184 @@
+        tempRelationships.add(REL_ORIGINAL);
+        tempRelationships.add(REL_FAILURE);
+        relationships = Collections.unmodifiableSet(tempRelationships);
+
+        final List<PropertyDescriptor> tempDescriptors = new ArrayList<>();
+        tempDescriptors.add(CLIENT_SERVICE);
+        tempDescriptors.add(QUERY);
+
+        propertyDescriptors = Collections.unmodifiableList(tempDescriptors);
+    }
+
+    protected ObjectMapper mapper = new ObjectMapper();
+
+    @Override
+    public Set<Relationship> getRelationships() {
+        return relationships;
+    }
+
+    @Override
+    public final List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return propertyDescriptors;
+    }
+
+    private volatile GraphClientService clientService;
+
+    @OnScheduled
+    public void onScheduled(final ProcessContext context) {
+        super.onScheduled(context);
+        clientService = context.getProperty(CLIENT_SERVICE).asControllerService(GraphClientService.class);
+    }
+
+    @Override
+    public void onTrigger(final ProcessContext context, final ProcessSession session) throws ProcessException {
+        FlowFile flowFile = session.get();
+
+        FlowFile output = flowFile != null ? session.create(flowFile) : session.create();
+        try (OutputStream os = session.write(output)) {
+            String query = getQuery(context, session, flowFile);
+            long startTimeMillis = System.currentTimeMillis();
+
+            os.write("[".getBytes());
+            Map<String, String> resultAttrs = clientService.executeQuery(query, getParameters(context, output), (record, hasMore) -> {
+                try {
+                    String obj = mapper.writeValueAsString(record);
+                    os.write(obj.getBytes());
+                    if (hasMore) {
+                        os.write(",".getBytes());
+                    }
+                } catch (Exception ex) {
+                    throw new ProcessException(ex);
+                }
+            });
+            os.write("]".getBytes());
+            os.close();
+
+            final long endTimeMillis = System.currentTimeMillis();
+
+            String executionTime = String.valueOf((endTimeMillis - startTimeMillis));
+            resultAttrs.put(EXECUTION_TIME, executionTime);
+            resultAttrs.put(CoreAttributes.MIME_TYPE.key(), "application/json");
+            output = session.putAllAttributes(output, resultAttrs);
+            session.transfer(output, REL_SUCCESS);
+            session.getProvenanceReporter().invokeRemoteProcess(output, clientService.getTransitUrl(),
+                String.format("The following query was executed in %s milliseconds: \"%s\"", executionTime, query)
+            );
+            if (flowFile != null) {
+                session.transfer(flowFile, REL_ORIGINAL);
+            }
+
+        } catch (Exception exception) {
+            getLogger().error("Failed to execute graph statement due to {}",
```

Comment:
If `executeQuery` throws an exception, the `output` flow file has been created, so it needs to be removed here.